image_writers_guide/guidelines.adoc

```diff
@@ -9,4 +9,135 @@
+
+Of course any downstream consumer who chooses to use "foo:latest" will be taking on the risk of any incompatible changes introduced.
+
+
+=== Do not start multiple processes
+It is not a good idea to start multiple services (like database, sshd, …​) inside one container. Containers are lighweight and can be easily linked together for multiprocess orchestration.  OpenShift allows you to easily colocate and comanage related images by grouping them into a single pod.  This will ensure they share a network namespace and storage for communication.  It also makes updates less disruptive as each image can be updated less frequently and independently.  Having a single process also makes signal handling flows clearer as you do not need to manage routing signals to spawned processes.
+
+
+=== Use exec in wrapper scripts
+See http://www.projectatomic.io/docs/docker-image-author-guidance/, subsection _Always `exec` in Wrapper Scripts_
+
+Also be aware that your process runs as PID 1 when running in a Docker container.  This means that if your main process terminates, the entire container will be stopped, killing any child processes you may have launched from your PID 1 process.  Additional implications to be aware of are documented http://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/[here].  For a deep dive on PID 1 and init systems, you can read https://felipec.wordpress.com/2013/11/04/init/[this] as well.
+
+
+=== Clean temporary files
+Every temporary files you create during the build process should be removed. This includes any files added with the ADD command too.  For example it is strongly recommended you run yum clean after performing yum install operations.  If you craft your `RUN` statement as follows, you an prevent the yum cache from ending up in an image layer:
+
+	RUN yum -y install mypackage && yum -y install myotherpackage && yum clean all -y
+
+Be aware that if you instead write:
+		
+		RUN yum -y install mypackage
+		RUN yum -y install myotherpackage && yum clean all -y
+
+Then the first `yum` invocation will leave extra files in that layer which cannot be removed by the later clean operation.  (The extra files will not be visible in the final image but they will be present in the underlying layers).
+
+Note: Although the current Docker build process does not allow shrinking the space used by the image when we remove something from an earlier layer by a command run in a later layer, this may change in the future.   This means that if you perform an `rm` command in a later layer, you will hide the files but you will not reduce the overall size of the image that needs to be downloaded.  Therefore, as with the yum clean example, it's best to remove files in the same command that created them, where possible, so they do not end up written to a layer.
+
```

Comment:
A note about chaining commands executed with `RUN` instruction may be good at this point. It's not only about cleaning up files, but also about reducing layers which make the image bit more complicated and it'll take more time to download and extract it too.